packages/next/vitals/vitals.ts

```diff
@@ -0,0 +1,29 @@
+import { useEffect, useRef } from 'react'
+import { NextWebVitalsMetric } from '../pages/_app'
+
+type ReportWebVitalsCallback = (webVitals: NextWebVitalsMetric) => any
+export const webVitalsCallbacks = new Set<ReportWebVitalsCallback>()
+const metrics: NextWebVitalsMetric[] = []
+
+export function trackWebVitalMetric(metric: NextWebVitalsMetric) {
+  metrics.push(metric)
+  webVitalsCallbacks.forEach((callback) => callback(metric))
+}
+
+export function useWebVitalsReport(callback: ReportWebVitalsCallback) {
+  const metricIndexRef = useRef(0)
+
+  useEffect(() => {
+    // Flush calculated metrics
+    const reportMetric = (metric: NextWebVitalsMetric) => callback(metric)
```

Comment:
```tsx
const reportMetric = (metric: NextWebVitalsMetric) => {
  callback(metric)
  metricIndexRef.current = metrics.length
}
```

the ref needs to be updated after future callbacks too, or those will be replayed.

Reply:
Good point! 👍